File: 프로그래머스/lv2/131127. 할인 행사/할인 행사.java
import java.util.*;
class Solution {
    public int solution(String[] want, int[] number, String[] discount) {
        int answer = 0;
        for(int i = 0; i < discount.length-9; i++) {
        	Map<String, Integer> wantMap = new HashMap<>();
        	for(int j = 0; j < want.length; j++) {
        		wantMap.put(want[j], number[j]);
        	}
        	for(int j = i; j < i + 10; j++) {
        		if(wantMap.containsKey(discount[j])) {
        			wantMap.put(discount[j], wantMap.get(discount[j])-1);
        		}
        	}
        	answer++;
        	for(String key : wantMap.keySet()) {
        		if(wantMap.get(key) != 0) {
        			answer--;
        			break;
        		}
        	}
        }
        return answer;
    }
}

File: 프로그래머스/2/131127. 할인 행사/할인 행사.js
function solution(want, number, discount) {
  const n = want.length
  const discountInfo = {}
  let res = 0
  let start = 0
  let end = 0
  while (end < discount.length) {
    discountInfo[discount[end]] = (discountInfo[discount[end]] ?? 0) + 1
    // 모두 살 수 있는지 체크
    if (end - start + 1 === 10) {
      let flag = true
      for (let i = 0; i < n; i++) {
        if ((discountInfo[want[i]] || 0) < number[i]) {
          flag = false
          break //불가능
        }
      }
      if (flag) res++
      discountInfo[discount[start++]]--
    }
    end++
  }
  return res
}